ui/packages/app/src/Blocks.svelte

```diff
@@ -256,19 +257,43 @@
 					outputs.every((v) => instance_map?.[v].instance) &&
 					inputs.every((v) => instance_map?.[v].instance)
 				) {
-					const req = fn({
-						action: "predict",
-						backend_fn,
-						frontend_fn,
-						payload: {
-							fn_index: i,
-							data: inputs.map((id) => instance_map[id].props.value)
-						},
-						queue: queue === null ? enable_queue : queue,
-						queue_callback: handle_update,
-						loading_status: loading_status,
-						cancels
-					});
+					cancels &&
+						cancels.forEach((fn_index) => {
+							prediction_map[fn_index].cancel(fn_index);
+						});
+
+					let payload = {
+						fn_index: i,
+						data: inputs.map((id) => instance_map[id].props.value)
+					};
+
+					if (frontend_fn) {
+						frontend_fn(
+							payload.data.concat(
+								outputs.map((id) => instance_map[id].props.value)
+							)
+						).then((v: []) => {
+							if (backend_fn) {
+								payload.data = v;
+								make_prediction();
+							} else {
+								handle_update({ data: v });
+							}
+						});
+					} else {
+						if (backend_fn) {
+							make_prediction();
+						}
+					}
+
+					function make_prediction() {
+						prediction_map[i] = app
+							.predict("/predict", payload)
+							.on("data", handle_update)
+							.on("status", (s) =>
+								loading_status.update({ ...s, fn_index: i })
+							);
+					}
```

Comment:
Not in love with this but will address it when I tackle the component loading + event handler refactoring.